src/bokeh/plotting/_renderer.py

```diff
@@ -116,6 +116,11 @@ def create_renderer(glyphclass, plot, **kwargs):
     hover_glyph = make_glyph(glyphclass, kwargs, hover_visuals)
     muted_glyph = make_glyph(glyphclass, kwargs, muted_visuals)
 
+    # Check if glyph has a specific _validate_with_data_source method, and if so call it.
+    for g in (glyph, nonselection_glyph, selection_glyph, hover_glyph, muted_glyph):
+        if callable(getattr(g, "_validate_with_data_source", None)):
+            g._validate_with_data_source(source)
```

Comment:
This could be made into a legitimate validator, barring any concerns about performance, if the code was moved to `glyph_renderer.py` and turned into a proper validator (i.e decorated with `@error`), while keeping the contextual validation function on glyphs, which should be also defined on the base class to avoid `hasattr`/`getattr`.